tensorflow/lite/kernels/irfft2d.cc

```diff
@@ -0,0 +1,426 @@
+  // for memory allocation. And it will be cast into double in Eval when being
+  // used.
+  fft_double_working_area->type = kTfLiteInt64;
+  // If fft_length is not a constant tensor, fft_double_working_area will be
+  // set to dynamic later in Prepare.
+  fft_double_working_area->allocation_type = kTfLiteArenaRw;
+
+  return kTfLiteOk;
+}
+
+TfLiteStatus ResizeOutputandTemporaryTensors(TfLiteContext* context,
+                                             TfLiteNode* node) {
+  const TfLiteTensor* input;
+  TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, kInputTensor, &input));
+  const int num_dims = NumDimensions(input);
+  TF_LITE_ENSURE(context, num_dims >= 2);
+  const TfLiteTensor* fft_length;
+  TF_LITE_ENSURE_OK(context,
+                    GetInputSafe(context, node, kFftLengthTensor, &fft_length));
+  const int32_t* fft_length_data = GetTensorData<int32_t>(fft_length);
+  // The lib, fft2d, can only handle fft_lengths of power of 2.
+  TF_LITE_ENSURE(context, IsPowerOfTwo(fft_length_data[0]));
+  TF_LITE_ENSURE(context, IsPowerOfTwo(fft_length_data[1]));
+
+  int fft_height, fft_width;
+  fft_height = fft_length_data[0];
+  fft_width = fft_length_data[1];
+  int fft_working_length = std::max(fft_height, fft_width / 2);
+  int half_fft_working_length = fft_working_length / 2;
+
+  // Resize output tensor.
+  TfLiteTensor* output;
+  TF_LITE_ENSURE_OK(context,
+                    GetOutputSafe(context, node, kOutputTensor, &output));
+  TfLiteIntArray* output_shape = TfLiteIntArrayCopy(input->dims);
+  output_shape->data[num_dims - 2] = fft_length_data[0];
+  output_shape->data[num_dims - 1] = fft_length_data[1];
+  TF_LITE_ENSURE_STATUS(context->ResizeTensor(context, output, output_shape));
+
+  // Resize temporary tensors, fft_integer_working_area.
+  TfLiteTensor* fft_integer_working_area;
+  TF_LITE_ENSURE_OK(
+      context, GetTemporarySafe(context, node, kFftIntegerWorkingAreaTensor,
+                                &fft_integer_working_area));
+  TfLiteIntArray* fft_integer_working_area_shape = TfLiteIntArrayCreate(1);
+  fft_integer_working_area_shape->data[0] =
+      2 + static_cast<int>(sqrt(fft_working_length));
+  TF_LITE_ENSURE_STATUS(context->ResizeTensor(context, fft_integer_working_area,
+                                              fft_integer_working_area_shape));
+
+  // Resize temporary tensors, fft_double_working_area.
+  TfLiteTensor* fft_double_working_area;
+  TF_LITE_ENSURE_OK(context,
+                    GetTemporarySafe(context, node, kFftDoubleWorkingAreaTensor,
+                                     &fft_double_working_area));
+  TfLiteIntArray* fft_double_working_area_shape = TfLiteIntArrayCreate(1);
+  fft_double_working_area_shape->data[0] =
+      half_fft_working_length + fft_width / 4;
+  TF_LITE_ENSURE_STATUS(context->ResizeTensor(context, fft_double_working_area,
+                                              fft_double_working_area_shape));
+
+  return kTfLiteOk;
+}
+
```

Comment:
Sorry that I forgot to submit my review a few days ago.

Reply:
Hey @lu-wang-g , sorry for the slow reply, and thanks for reviewing.

Yes, I can absolutely do that. Some of the functions overlap with 95% of code but not 100%. I'll have a look which of those I'll move into a common function that takes an additional forward/backward mode parameter, and which ones I deem easier to read when left separate. But you're right, there's a decent chunk that can go into a common file.

Some questions:
- I would name that file `rfft2d_shared.h`, similar to `lstm_shared.h`. Is that ok?
- Right now, I have two namespaces: `tflite::ops::builtin::rfft2d` and `tflite::ops::custom::irfft2d`. Is it ok to put the shared functionality into the `tflite::ops::builtin::rfft2d` namespace, or should I choose something else?

Reply:
It seems I may need a bit of help with this. I tried implementing the proposal of my previous comment, but get linker errors when running ` bazel test //tensorflow/lite/kernels:rfft2d_test`:

```
dyld: Symbol not found: __ZN6tflite3ops7builtin6rfft2dL20InitTemporaryTensorsEP13TfLiteContextP10TfLiteNode
  Referenced from: /private/var/tmp/_bazel_thomas/deaa089c7e830773935e4550ab22fa14/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/lite/kernels/rfft2d_test.runfiles/org_tensorflow/tensorflow/lite/kernels/../../../_solib_darwin_x86_64//libtensorflow_Slite_Skernels_Slibbuiltin_Uop_Ukernels.so
  Expected in: flat namespace
```

Which is surprising to me given that I defined the following library in the `BUILD` file:

```
cc_library(
    name = "rfft2d_shared",
    srcs = ["rfft2d_shared.cc"],
    hdrs = ["rfft2d_shared.h"],
    compatible_with = get_compatible_with_portable(),
    copts = tflite_copts(),
    deps = [
        ":kernel_util",
        "//tensorflow/lite/kernels/internal:tensor",
    ],
)
```

And then listed this as a dependency for the builtin_ops and custom_ops target.

BUILD file, rfft2d_shared.h/cc are here: https://gist.github.com/tcwalther/44a7a9884b5bb21b4e498c885eb21be6

I also still wonder whether it may be worth putting everything into a single file (i.e., everything in rfft2d.cc, getting rid of irfft2d.cc entirely). I'm just not 100% sure about it. The code of these two files is very similar, but differs in small spaces, and unifying it with several if/else statements here and there might make it a lot less readable for someone trying to understand what's going on.

If you could give me some advice on how to proceed on this, I'd greatly appreciate it!

Reply:
Please note that if I do put everything into rfft2d.cc, this would make the IRFFT a built-in operator, too. Which would be perfect IMO, but I just want to point it out instead of sneaking it in.

Reply:
I've created an attempt at putting everything into `rfft2d.cc` to minimize code duplication. As said before, I'm not sure it's more readable:

https://gist.github.com/tcwalther/ed573256748cfc48847a4304068d5f69

I don't understand yet how to register a new builtin op and properly add it to the schema; as such, the tests currently fail with this file. If you want, you can look at this in approach its own branch at https://github.com/tcwalther/tensorflow/tree/tw-tflite-irfft-one-file

Please let me know which approach you prefer.

Reply:
+1 I prefer this approach to put everything in rfft2d.cc (maybe we should change the file name).

Reply:
Thanks @thaink. In that case, could you please give me guidance on how to add a built-in op properly to the schema? I've had a look at the git commit that promotes RFFT2D to built-in: https://github.com/tensorflow/tensorflow/commit/bc295ed3bc729e94a0aa16214ae3c5a8b642858c

It seems to me that some of these changes were manual where others seem to be the result of some script that one runs (in particular, `schema_generated.h`). If could point me to a documentation outlining the required steps, I will update this PR accordingly.

Reply:
Oh, I see. @renjie-liu Is it acceptable to define both ops in the same file but leave one as custom op?